pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -88,6 +81,10 @@ func (cfg *Config) Validate() error {
 	return nil
 }
 
+func cardinalityBasedShardingEnabled(cfg *Config) bool {
```

Comment:
Looks like this should be a method on `*Config`.

Reply:
fixed in https://github.com/grafana/mimir/pull/4254/commits/e979127ddb68d035cfe6c3cf52cd2fc47a436f30